authentication/ldap_authentication_provider.py

```diff
@@ -12,35 +12,147 @@
-    def __init__(self, directory_url):
-        self.directory_url = directory_url
+    def __init__(self,
+                 ldap_url,
+                 base_dn=None,
+                 admin_user_id=None,
+                 admin_password=None,
+                 user_id_attribute=None):
+        self.lgr = logging.getLogger(constants.FLASK_SECUREST_LOGGER_NAME)
+        self.ldap_url = ldap_url
+        self.base_dn = base_dn
+        self.admin_user_id = admin_user_id
+        self.admin_password = admin_password
+        self.user_id_attribute = user_id_attribute
 
+    # LDAP authentication can be done by subtree match or according to
+    # dn attribute.
     def authenticate(self, userstore=None):
-        username, password = utils.get_basic_http_authentication_info()
+        who, password = self.get_creds_from_request()
+        if self.is_subtree_authentication_method():
+            self.authenticate_by_attribute(who,
+                                           password,
+                                           default_attr=DEFAULT_AUTH_ATTR_NAME)
+        elif self.is_dn_authentication_method():
+            self.try_bind_s(who, password)
+        else:
+            raise RuntimeError('Failed determining authentication method. '
+                               'LDAP configuration is invalid.')
+        return who
+
+    def authenticate_by_attribute(self, who, password, default_attr=None):
+        try:
+            conn = self.synchronous_bind(self.admin_user_id,
+                                         self.admin_password)
+        except AuthenticationException as e:
+            err = 'Failed binding to \'{0}\' using the dedicated ' \
+                  'admin_dn: {1}, provided in the configuration; {2}'\
+                .format(self.ldap_url, who, e)
+            self.lgr.warning(err)
+            raise RuntimeError(err)
+
+        attribute_name = self.user_id_attribute or default_attr
+        search_filter = '{0}={1}'.format(attribute_name, who)
+
+        result_list = []
+        try:
+            # Search for matches under the base_dn subtree using the attribute
+            # search filter.
+            ldap_result_id = conn.search(self.base_dn,
+                                         ldap.SCOPE_SUBTREE,
+                                         search_filter,
+                                         [DN_ATTRIBUTE_NAME])
+        except ldap.LDAPError as e:
+            err = 'Failed searching for user under base DN: {0}, according ' \
+                  'to the following attribute filter pattern {1}; {2}'\
+                .format(self.base_dn, search_filter, e)
+            self.handle_authentication_error(err, conn)
+
+        try:
+            # Retrieve query results according to the result ID in order to
+            # obtain the user DN to bind with.
+            while True:
+                result_type, result_data = conn.result(ldap_result_id, 0)
```

Comment:
1. It is indeed surprising but by default, UID is not set to be unique on LDAP (http://stackoverflow.com/questions/13747795/creating-user-accounts-with-unique-uid-in-openldap). There is however an option to configure LDAP so that it would be unique. Regarding the fact that we can raise an error way faster, totally true.
2. Couldn't find anything. What's wrong with search though?